main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala

```diff
@@ -248,47 +248,47 @@ object TypeConstructor {
+    def kind: Kind = Kind.JvmConstructorOrMethod
   }
 
   /**
    * A type constructor that represents the type of a Java method.
    */
   case class JvmMethod(method: Method) extends TypeConstructor {
-    def kind: Kind = Kind.Star
+    def kind: Kind = Kind.JvmConstructorOrMethod
   }
 
   /**
-   * A type constructor that represents the _return type_ of a Java method.
+   * A type constructor that represents the _return type_ of a Java static method.
    *
-   * A method return type can be resolved when the receiver object and argument types are known.
+   * A static method return type can be resolved when the argument types are known.
    *
    * A few examples:
    *
-   * - The type: `Apply(InvokeMethod("length", 0), String)` is equivalent to `Int32`.
-   * - The type: `Apply(Apply(InvokeMethod("startsWith", 1), String), String)` is equivalent to `Bool`.
-   * - The type: `Apply(Apply(Apply(InvokeMethod("substring", 2), String), Int32), Int32)` is equivalent to `String`.
+   * - The type: `Apply(InvokeStaticMethod(String.class, "valueOf", 1), Bool)` is equivalent to `String`.
+   * - The type: `Apply(InvokeStaticMethod(String.class, "valueOf", 1), Char)` is equivalent to `String`.
    *
-   * The type constructor requires at least one type argument: the type of the receiver object.
+   * The type constructor does not require any arguments.
    */
-  case class MethodReturnType(methodName: String, arity: Int) extends TypeConstructor {
-    def kind: Kind = Kind.mkArrow(List(Kind.JvmConstructorOrMethod, Kind.Star))
+  case class StaticMethodReturnType(clazz: Class[_], name: String, arity: Int) extends TypeConstructor {
+    def kind: Kind = Kind.mkArrow(arity)
   }
 
   /**
-   * A type constructor that represents the _return type_ of a Java static method.
+   * A type constructor that represents the _return type_ of a Java method.
    *
-   * A static method return type can be resolved when the argument types are known.
+   * A method return type can be resolved when the receiver object and argument types are known.
    *
    * A few examples:
    *
-   * - The type: `Apply(InvokeStaticMethod(String.class, "valueOf", 1), Bool)` is equivalent to `String`.
-   * - The type: `Apply(InvokeStaticMethod(String.class, "valueOf", 1), Char)` is equivalent to `String`.
+   * - The type: `Apply(InvokeMethod("length", 0), String)` is equivalent to `Int32`.
+   * - The type: `Apply(Apply(InvokeMethod("startsWith", 1), String), String)` is equivalent to `Bool`.
+   * - The type: `Apply(Apply(Apply(InvokeMethod("substring", 2), String), Int32), Int32)` is equivalent to `String`.
    *
-   * The type constructor does not require any arguments.
+   * The type constructor requires at least one type argument: the type of the receiver object.
    */
-  case class StaticMethodReturnType(clazz: Class[_], name: String, arity: Int) extends TypeConstructor {
-    def kind: Kind = Kind.mkArrow(arity)
+  case object MethodReturnType extends TypeConstructor {
```

Comment:
Let's add `@EliminatedBy` tags to these to mark that they are eliminated by the `TypeReconstruction` phase